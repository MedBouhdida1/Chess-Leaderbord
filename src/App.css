.App {
  font-family: sans-serif;
  text-align: center;
}

.pagination {
  display: flex;
  justify-content: center;
  margin-top: 20px;
}

.page-item {
  display: inline-block;
  margin: 0 2px;
}

.page-link {
  padding: 8px 12px;
  border: 1px solid #dee2e6;
  color: #333;
  cursor: pointer;
}

.page-link:hover {
  background-color: #f8f9fa;
}

.page-link:focus {
  box-shadow: none;
}

.page-item.disabled .page-link {
  opacity: 0.6;
  pointer-events: none;
  cursor: not-allowed;
}

.page-item.active .page-link {
  background-color: #007bff;
  border-color: #007bff;
  color: #fff;
}