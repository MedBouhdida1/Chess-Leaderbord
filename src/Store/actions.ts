export const FETCH_DATA_REQUEST = 'FETCH_DATA_REQUEST';
export const FETCH_DATA_SUCCESS = 'FETCH_DATA_SUCCESS';
export const FETCH_DATA_FAILURE = 'FETCH_DATA_FAILURE';
export const DELETE_PLAYER = 'DELETE_PLAYER';
export const ADD_PLAYER = 'ADD_PLAYER';
export const EDIT_PLAYER = 'EDIT_PLAYER';

export const fetchDataRequest = () => ({
  type: FETCH_DATA_REQUEST,
});

export const fetchDataSuccess = (data:any) => ({
  type: FETCH_DATA_SUCCESS,
  payload: data,
});

export const fetchDataFailure = (error:any) => ({
  type: FETCH_DATA_FAILURE,
  payload: error,
});

export const removePlayer = (id:number) => ({
  type: DELETE_PLAYER,
  payload: id,
});
export const addPlayer = (player:any) => ({
  type: ADD_PLAYER,
  payload: player,
});
export const editPlayer = (player_id:number,updatedPlayerData:any) => ({
  type: EDIT_PLAYER,
  payload: {
    player_id,
    updatedPlayerData,
  },
});