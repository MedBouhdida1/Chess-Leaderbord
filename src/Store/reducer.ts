import { FETCH_DATA_REQUEST, FETCH_DATA_SUCCESS, FETCH_DATA_FAILURE, DELETE_PLAYER, ADD_PLAYER, EDIT_PLAYER } from './actions';

const initialState = {
  data: [],
  loading: false,
  error: null,
};

export const playerReducer = (state = initialState, action:any) => {

    switch (action.type) {
      case FETCH_DATA_REQUEST:
        return {
          ...state,
          loading: true,
        };
  
      case FETCH_DATA_SUCCESS:
        return {
          ...state,
          loading: false,
          data: action.payload, 
          error: null,
        };
  
      case FETCH_DATA_FAILURE:
        return {
          ...state,
          loading: false,
          error: action.payload,
        };
      case DELETE_PLAYER:
        return {
          ...state,
          loading: false,
          data: Array.isArray(state.data) ? state.data.filter((p:any) => p.player_id !== action.payload) : [],
          error: null,
        };
        case ADD_PLAYER:
          return {
            ...state,
            loading: false,
            data: Array.isArray(state.data) ? [action.payload, ...state.data] : [action.payload],
            error: null,
          };
          case EDIT_PLAYER: {
            const { player_id, updatedPlayerData } = action.payload;
            const playerIndex = state.data.findIndex((player:any) => player.player_id === player_id);
            if (playerIndex !== -1) {
              const newData:any = [...state.data];
              newData[playerIndex] = {
                ...newData[playerIndex],
                ...updatedPlayerData,
              };
          
              return {
                ...state,
                loading: false,
                data: newData,
                error: null,
              };
            }
          break;
          }
      default:
        return state;
    }
  };