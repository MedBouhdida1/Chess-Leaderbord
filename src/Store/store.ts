import { configureStore } from "@reduxjs/toolkit";
import { playerReducer } from "./reducer";





export const store = configureStore({
    reducer: {
        players:playerReducer
    },
})