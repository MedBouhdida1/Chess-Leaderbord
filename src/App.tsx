import "./App.css";
import LoadingSpinner from "./LoadingSpinner";

import List from "./Components/List";
import { useToast } from '@chakra-ui/react'
import { useDispatch, useSelector } from "react-redux";
import { useEffect } from "react";
import { FETCH_DATA_FAILURE, FETCH_DATA_REQUEST, FETCH_DATA_SUCCESS, fetchDataFailure, fetchDataRequest, fetchDataSuccess } from "./Store/actions";
import { store } from "./Store/store";
import axios from "axios";


//Api Endpoint
const API_ENDPOINT = "https://api.chess.com/pub/leaderboards";

const App: React.FC = () => {






  //redux

  const dispatch = useDispatch();


  const { data: test, loading, error } = useSelector((state: any) => state.players);

  const fetchDataFromApi = async () => {
    dispatch(fetchDataRequest());

    try {
      const response = await axios.get(API_ENDPOINT);
      const data = response.data;
      console.log("data ", data)

      dispatch(fetchDataSuccess(data.daily));

      console.log("store content", store.getState().players)
    } catch (error: any) {
      dispatch(fetchDataFailure(error?.message));
    }
  };

  useEffect(() => {

    fetchDataFromApi();

  }, [dispatch]);




  return (
    <>
      <div className="container">
        <h1 className="d-flex justify-content-center ">Chess Leaderboards</h1>


        <hr />

        {/* If the error variable is truthy (i.e., not null, undefined, or false), the code renders a <p> element with the text "Something went wrong ...". */}
        {error && <p data-testid="error">Something went wrong ...</p>}

        {/* If the isLoading variable is true, indicating that data is being loaded, the code renders a <LoadingSpinner /> component. */}
        {/* If the isLoading variable is false, indicating that data has finished loading, the code renders a <List> component. */}
        {loading ? (
          <LoadingSpinner />
        ) : (
          <List />
        )}
      </div>

    </>
  );
};



export default App;
