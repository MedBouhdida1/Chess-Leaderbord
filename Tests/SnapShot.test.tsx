
import { describe, it, expect, vi } from "vitest";
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from "react-redux";
import { store } from "../src/Store/store";
import App from "../src/App";
import React from "react";

//***********************************SnapShot  TEST *********************************** والله أعلم*/
describe('List of Player', () => {
    it('renders snapshot', () => {
        const { container } = render(
            <Provider store={store}>
                <App />
            </Provider>);
        expect(container.firstChild).toMatchSnapshot();
    });
});
