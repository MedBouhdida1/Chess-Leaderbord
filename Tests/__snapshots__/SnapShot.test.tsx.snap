// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List of Player > renders snapshot 1`] = `
<div
  class="container"
>
  <h1
    class="d-flex justify-content-center "
  >
    Chess Leaderboards
  </h1>
  <hr />
  <div
    class="spinner-container"
    data-testid="loading"
  >
    <div
      class="loading-spinner"
    />
  </div>
</div>
`;
